### ISSUE TITLE
SQL Server NVARCHAR literal is not rendered correctly

### ISSUE THREAD

**Issue by lyginne** [jOOQ/jOOQ#9540]
### Expected behavior and actual behavior:

```scala
String result = DSL.using(SQLDialect.SQLSERVER).renderInlined(`val`("йцукен", SQLDataType.NVARCHAR))
```

#### expected value: 
`N'йцукен'`
#### actual value: 
`'йцукен'`
#### this test should be passed:

```java
assertEquals("N'йцукен'", result, "SQLSERVER dialect: nvarchar value vas not rendered properly")
```

### Versions:

- jOOQ:3.12.3
- Java:11
- Database (include vendor):MSSQL 
- OS: Mac os catalina

**Comment by lukaseder** [comment]
Thank you very much for your report. jOOQ currently doesn't implement a dedicated data type `Binding` for the `NVARCHAR` type (and other similar JDBC types like `NCLOB` or `NCHAR`). This is because the default `Binding` implementation is based on a `Class<T>` reference (e.g. `String.class`), not a `DataType<T>` reference (e.g. `SQLDataType.NVARCHAR`).

The workaround right now is to apply your own custom data type binding to all relevant columns in generated code:
https://www.jooq.org/doc/latest/manual/sql-building/queryparts/custom-bindings

Once you have such a binding, you can use it also for `val()` calls:

```java
Field<String> value = val("йцукен", SQLDataType.NVARCHAR.asConvertedDataType(new MyBinding());
String result = DSL.using(SQLDialect.SQLSERVER).renderInlined(value);
```

**Comment by lukaseder** [comment]
Related issue: https://github.com/jOOQ/jOOQ/issues/9548

**Comment by lukaseder** [comment]
Finally fixing this now. This depends on the availability of a `DataType<T>` reference in `DefaultBinding`, which was implemented in jOOQ 3.14 via #10362. Without it, we cannot decide whether a `Binding<String, ?>` should bind a `String` or `NString`

**Comment by lukaseder** [comment]
I'll add a `DataType.isNString()` method to recognise such data types: https://github.com/jOOQ/jOOQ/issues/10368

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.14. This depends on a variety of recent internal improvements around `DataType` implementations, which is why this fix cannot be backported to older versions.
